File: writeups/35_Search_Insert_Position/ruby/main.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def search_insert(nums, target)
  head = 0
  tail = nums.length - 1

  while head <= tail
    center = (head + tail) / 2
    if nums[center] == target
      return center
    elsif nums[center] < target
      head = center + 1
    else
      tail = center - 1
    end
  end

  head
end

File: writeups/35_Search_Insert_Position/typescript/main.ts
function searchInsert(nums: number[], target: number): number {
  let head: number = 0;
  let tail = nums.length - 1;
  while (head <= tail) {
    const center = Math.floor((head + tail) / 2);
    if (nums[center] === target) {
      return center;
    } else if (nums[center] < target) {
      head = center + 1;
    } else {
      tail = center - 1;
    }
  }

  return head;
}
